zero/rect_grid.c

```diff
@@ -22,6 +24,138 @@ gkyl_rect_grid_init(struct gkyl_rect_grid *grid, int ndim,
   }
 }
 
+GKYL_CU_DH
+void gkyl_rect_grid_find_cell(struct gkyl_rect_grid *grid, const double *point, bool pickLower, const int **knownIdx, int *cellIdx){
+  int nDim = grid->ndim;
+  int searchNum = 0;
+  int searchDim[GKYL_MAX_DIM], *idxOut;
+  int *dimTrans[GKYL_MAX_DIM];
+  int plusminus[2] = {-1,1};
+  bool allLessEq = true;
+  int iStart[GKYL_MAX_DIM], iEnd[GKYL_MAX_DIM], iMid[GKYL_MAX_DIM], iNew[GKYL_MAX_DIM];
+  int *cells, lowHighCellIdx[2*GKYL_MAX_DIM];
+  double lowerInDir[nDim], upperInDir[nDim];
+  double low, high;
+  
+  for(int d=0; d<nDim; d++){
+    dimTrans[d] = (int*)malloc(1*sizeof(int));
+    if(knownIdx[d]==NULL){
+      searchDim[searchNum] = d;
+      *dimTrans[d] = searchNum;
+      searchNum = searchNum + 1;
+    }else{
+      dimTrans[d] = NULL;      
+      cellIdx[d] = *knownIdx[d];
+      low = grid->lower[d]+(*knownIdx[d]-1)*grid->dx[d];
+      high = grid->lower[d]+(*knownIdx[d])*grid->dx[d];
+      assert(low<point[d]&&high>point[d]);
+    }
+  }
+
+  cells = grid -> cells;
+  for(int d=0; d<searchNum; d++){
+    iStart[d] = 1;
+    iEnd[d] = cells[searchDim[d]];
+    iMid[d] = 0;
+    iNew[d] = 0;
+  }
+  /* Below we use a binary search. That is, if the i-th coordinate of the point in
+   * question is below(above) the ith-coordinate of the current mid-point, we search
+   * the lower(upper) half along that direction in the next iteration.
+   */
+  while (allLessEq){
+    for(int d=0; d<searchNum; d++){
```

Comment:
Here and elsewhere:
- please remove the curly brackets in for loops that contain a single instruction.
- Leave at least 2 spaces between code and comment.
- Leave 1 space after // in your comment.

Reply:
Should the curly brackets be neglected from the outer for loop in double for loops even when the inner for loop has multiple instructions (assuming the outer for loop only contains the inner for loop), such as the following?

for (int i=1; i<a; i++) {  // this curly bracket
   for (int j=1; j<b; j++) {
      x=i+j;
      .
      .
      .
   }
}

Reply:
I think not, good catch. An exception to my first bullet above is when that single instruction is another loop or switch